### ISSUE TITLE
TT-SSM

### ISSUE THREAD

**Issue by leyuan-sun** [Chen-Yang-Liu/RSCaMa#1]
Thansk for your work.
I cannot find the implementation of moduel TT-SSM. Especially, how to convert It1 ={I11,I12,I13}, It2={I21, I22, I23}  to {I11,I21, I12, I22, I13,I23}?

**Comment by leyuan-sun** [comment]
Q1: does camalayer is the same with fuselayer?
 for i in range(n_layers):
            self.CaMalayer_list.append(nn.ModuleList([
                CaMambaModel(config_1),
                CaMambaModel(config_1),
            ]))
            self.fuselayer_list.append(nn.ModuleList([
                CaMambaModel(config_2),
                CaMambaModel(config_2),
            ]))

Q2: https://github.com/Chen-Yang-Liu/RSCaMa/blob/3d32571990aafd6802a731b757d547acf698f10b/model/model_encoder_attMamba.py#L253 I think this line is the fusion of two images(TT-SSM?), but I did not see the rearrange interleaving manner as you introduced in the paper?

Q3: this part is TT-SSM? since you comment the bitemporal fusion in the code, but I do not think so, there is even no mamba block in this part
            # # bitemporal fusion
            if i == self.n_layers-1:
                img1_cap = img_sa1.transpose(-1, 1).view(batch, c, h, w)
                img2_cap = img_sa2.transpose(-1, 1).view(batch, c, h, w)
                feat_cap = torch.cat([img1_cap, img2_cap], dim=1)
                feat_cap = self.LN_list[i](self.Conv1_list[i](feat_cap))
                # feat_cap = self.Conv1_list[i](feat_cap)
                img_fuse = feat_cap.view(batch, c, -1).transpose(-1, 1)#.unsqueeze(-1) # (batch_size, L, D)
                img_fuse = self.LN_norm(img_fuse).unsqueeze(-1)
                img_list.append(img_fuse)

**Comment by leyuan-sun** [comment]
Sorry for many questions, if I were wrong, plz let me know, thanks ahead!!

**Comment by Chen-Yang-Liu** [comment]
> Thansk for your work. I cannot find the implementation of moduel TT-SSM. Especially, how to convert It1 ={I11,I12,I13}, It2={I21, I22, I23} to {I11,I21, I12, I22, I13,I23}?

Please see the updated code, I have added comments as follows:
![image](https://github.com/Chen-Yang-Liu/RSCaMa/assets/77491712/67c6eee0-9364-4193-8388-8dfcac831a3e)

**Comment by leyuan-sun** [comment]
Thank you very much for reply. I see. 
I have one more question about mamba if you could answer I will be appreciate.
are you using the output of the last token in the seuqence to the decoder? why in Vim people using the class token(in the middle) as the output?

**Comment by Chen-Yang-Liu** [comment]
> Thank you very much for reply. I see. I have one more question about mamba if you could answer I will be appreciate. are you using the output of the last token in the seuqence to the decoder? why in Vim people using the class token(in the middle) as the output?

Our task is different from the image classification task in VIM.
Instead of using the last token as the decoder input, we transform the bi-temporal sequence processed by CaMa as the decoder input. Please refer to our paper for details: https://arxiv.org/pdf/2404.18895

**Comment by leyuan-sun** [comment]
Thanks for your reply, I have read your paper carefully. So the output of CaMa is the output of the last element in your input sequence, right? such as [It1, It2 ... Itn] is the input to CaMa, and the output is the output of Itn, right?

**Comment by Chen-Yang-Liu** [comment]
> Thanks for your reply, I have read your paper carefully. So the output of CaMa is the output of the last element in your input sequence, right? such as [It1, It2 ... Itn] is the input to CaMa, and the output is the output of Itn, right?

The input to the decoder is not a token (Itn) but a sequence of tokens. Please refer to formula (7) and the public code.

**Comment by leyuan-sun** [comment]
I see, you used all the sequence outputs (B,L,D)instead of using cls_token's (B,D) in Vim, right? Btw, do you try to add the pos_embed in your framework as Vim?

**Comment by Chen-Yang-Liu** [comment]
> I see, you used all the sequence outputs (B,L,D)instead of using cls_token's (B,D) in Vim, right? Btw, do you try to add the pos_embed in your framework as Vim?

yes, Figure 1 and Formula (4) show the posion embedding:  https://arxiv.org/pdf/2404.18895
